lib/libc/audio/lib_buffer.c

```diff
@@ -0,0 +1,208 @@
+#include <debug.h>
+
+#include <tinyara/audio/audio.h>
+#include "libc.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Configuration ************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: apb_semtake
+ *
+ *       Take an Audio Pipeline Buffer.
+ *
+ ****************************************************************************/
+
+static void apb_semtake(FAR struct ap_buffer_s *apb)
+{
+	/* Take the semaphore (perhaps waiting) */
+
+	while (sem_wait(&apb->sem) != 0) {
+		/* The only case that an error should occr here is if
+		 * the wait was awakened by a signal.
+		 */
+
+		ASSERT(errno == EINTR);
+	}
+}
+
+/****************************************************************************
+ * Name: apb_semgive
+ ****************************************************************************/
+
+#define apb_semgive(b) sem_post(&b->sem)
+
+/****************************************************************************
+ * Name: apb_alloc
+ *
+ * Allocate an Audio Pipeline Buffer for use in the Audio sub-system.  This
+ * will perform the actual allocate based on buffer data format, number of
+ * channels, etc. and prepare the buffer for consumption.
+ *
+ ****************************************************************************/
+
+int apb_alloc(FAR struct audio_buf_desc_s *bufdesc)
+{
+	uint32_t bufsize;
+	int ret;
+	struct ap_buffer_s *apb;
+
+	DEBUGASSERT(bufdesc->u.ppBuffer != NULL);
+
+	/* Perform a user mode allocation */
+
+	bufsize = sizeof(struct ap_buffer_s) + bufdesc->numbytes;
+	apb = lib_umalloc(bufsize);
+	*bufdesc->u.ppBuffer = apb;
+
+	/* Test if the allocation was successful or not */
+
+	if (*bufdesc->u.ppBuffer == NULL) {
+		ret = -ENOMEM;
+	} else {
+		/* Populate the buffer contents */
+
+		memset(apb, 0, bufsize);
+		apb->i.channels = 1;
+		apb->crefs = 1;
+		apb->nmaxbytes = bufdesc->numbytes;
+		apb->nbytes = 0;
+		apb->flags = 0;
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		apb->session = bufdesc->session;
+#endif
+
+		sem_init(&apb->sem, 0, 1);
+		ret = sizeof(struct audio_buf_desc_s);
```

Comment:
Are you sure, it should return the size of struct ?   I believe it should return the size of allocated memory ( bufsize )
Kindly cross check once